sotodlib/wiregrid/calibrate.py

```diff
@@ -0,0 +1,404 @@
+#!/usr/bin/python3
+
+# import modules
+# python standard
+import os
+import sys
+import scipy
+from scipy import odr
+import numpy as np
+from iminuit import cost, Minuit
+from matplotlib.patches import Ellipse, Circle
+
+# SO specific
+from sotodlib import core
+from sotodlib.core import Context
+from sotodlib.io import hk_utils
+import sotodlib.io.load_smurf as ls
+from sotodlib.io.load_smurf import Observations
+
+wg_degpercount = 360/52000
+stream_ids = ['ufm_mv19', 'ufm_mv18', 'ufm_mv22', 'ufm_mv29', 'ufm_mv7', 'ufm_mv9', 'ufm_mv15'] # e.g. MF1
+
+def get_det_angle(aman, **kwargs):
+    """
+    Calibrate TODs and get det_angle_raw by using the wire grid.
+    Before using this code, your axis manager must have two field implemented by HWP, i.e., demodQ and demodU
+
+    Parameters:
+    - aman : AxisManager
+    - wg_config : yaml file for the settings to load the wire grid meta data
+        e.g. wg_config = '/so/home/hnakata/scratch/metadata/context/2401_wg_satp1.yaml'
+    - wg_offset : the offset angle of wires defined by the hardware of the Grid Loader
+    - stopped_time : the stopped time of your target operation
+    Return:
+    - aman : wraped with the calibration data of the wire grid. Fields are explained in each method.
+    """
+    _config = kwargs['wg_config']
+    _offset = kwargs['wg_offset']
+    _stopped_time = kwargs['stopped_time']
+    #
+    # wrap the house-keeping data of the wire grid
+    aman = wg_wrap_hk(aman, _config)
+    #
+    # restrict AxisManger within the operation range
+    idx_wg_inside = _get_operation_range(aman)
+    aman = aman.restrict('samps', (idx_wg_inside[0], idx_wg_inside[-1]), in_place=False)
+    #
+    # wrap the demodQ/demodU related with the steps of the wire grid
+    # AxisManager here must have demodQ/demodU by the HWP
+    aman = wg_wrap_QU(aman, _stopped_time)
+    #
+    # fit the QU signal created by the wire grid
+    aman = wg_get_cfitres(aman)
+    aman = wg_get_lfitres(aman)
+    #
+    # get the det angles
+
+    det_angle_raw = 0.5*(aman.wg.lfitval[:,1] + aman.wg.lfitval[:,0]*(_offset))%180
+    det_angle_raw_err = 0.5*aman.wg.lfiterr[:,1] # temporal value
+    #
+    aman.wg.wrap('det_angle_raw', det_angle_raw, [(0, 'dets')])
+    aman.wg.wrap('det_angle_raw_err', det_angle_raw_err, [(0, 'dets')])
+    #
+    return aman
+
+# wrap HouseKeeping data
+def wg_wrap_hk(aman, wg_config):
```

Comment:
I think call wg_config `ctx` or `wg_ctx`.

Reply:
Okay, i'll follow the customs.